eth2-ssv/node/node.go

```diff
@@ -0,0 +1,243 @@
+								n.logger.Error("Failed to deserialize pubkey from duty")
+							}
+							v := n.validatorsMap[pubKey.SerializeToHexStr()]
+							logger.Info("starting duty processing start for slot")
+							go v.ExecuteDuty(n.context, slot, duty)
+						} else {
+							logger.Info("scheduling duty processing start for slot")
+							if err := n.slotQueue.Schedule(duty.PublicKey, slot, duty); err != nil {
+								n.logger.Error("failed to schedule slot")
+							}
+						}
+					}(slot)
+				}
+			}(duty)
+		}
+	}
+}
+
+// setupValidators for each validatorShare with proper ibft wrappers
+func (n *SsvNode) setupValidators() {
+	validatorShares, err := n.validatorStorage.GetAllValidatorShares()
+	if err != nil {
+		n.logger.Fatal("Failed to get all validator shares", zap.Error(err))
+	}
+
+	res := make(map[string]*validator.Validator)
+	for _, validatorShare := range validatorShares {
+		res[validatorShare.ValidatorPK.SerializeToHexStr()] = validator.New(n.context, n.logger, validatorShare, n.ibftStorage, n.network, n.ethNetwork, n.beacon, validator.Options{
+			SlotQueue:                  n.slotQueue,
+			SignatureCollectionTimeout: n.signatureCollectionTimeout,
+		})
+	}
+	n.logger.Info("setup validators done successfully", zap.Int("count", len(res)))
+	n.validatorsMap = res
+}
+
+// startValidators functions (queue streaming, msgQueue listen, etc)
+func (n *SsvNode) startValidators() {
+	for _, v := range n.validatorsMap {
+		if err := v.Start(); err != nil {
+			n.logger.Error("failed to start validator", zap.Error(err))
+			continue
+		}
+	}
+}
+
+// startStreamDuties start to stream duties from the beacon chain
+func (n *SsvNode) startStreamDuties() {
+	var pubKeys [][]byte
+	var err error
+	for _, val := range n.validatorsMap {
+		pubKeys = append(pubKeys, val.ValidatorShare.ValidatorPK.Serialize())
+	}
+	n.streamDuties, err = n.beacon.StreamDuties(n.context, pubKeys)
+	if err != nil {
+		n.logger.Error("failed to open duties stream", zap.Error(err))
+	}
+	if n.pubkeysUpdateChan == nil {
+		n.pubkeysUpdateChan = make(chan bool) // first init
+	} else {
+		n.pubkeysUpdateChan <- true // update stream duty listener in order to fetch newly added pubkeys
+	}
+
+	n.logger.Info("start streaming duties")
+}
+
+// getSlotStartTime returns the start time for the given slot
+func (n *SsvNode) getSlotStartTime(slot uint64) time.Time {
+	timeSinceGenesisStart := slot * uint64(n.ethNetwork.SlotDurationSec().Seconds())
+	start := time.Unix(int64(n.ethNetwork.MinGenesisTime()+timeSinceGenesisStart), 0)
+	return start
+}
+
+// getCurrentSlot returns the current beacon node slot
+func (n *SsvNode) getCurrentSlot() int64 {
+	genesisTime := int64(n.ethNetwork.MinGenesisTime())
+	currentTime := time.Now().Unix()
+	return (currentTime - genesisTime) / 12
+}
+
+// getEpochFirstSlot returns the beacon node first slot in epoch
+func (n *SsvNode) getEpochFirstSlot(epoch uint64) uint64 {
+	return epoch * 32
```

Comment:
Recommend parametrizing the 32 to make it changeable in different config settings